File: negotiation-cordapp/workflows-java/src/test/java/negotiation/flows/ProposalFlowTests.java
package negotiation.flows;

import com.google.common.collect.ImmutableList;
import negotiation.states.ProposalState;
import net.corda.core.contracts.StateAndRef;
import net.corda.core.identity.Party;
import org.junit.Assert;
import org.junit.Test;

import java.util.List;
import java.util.concurrent.ExecutionException;

public class ProposalFlowTests extends FlowTestsBase {

    @Test
    public void proposalFlowCreatesTheCorrectProposalsInBothNodesVaultsWhenInitiatorIsBuyer() throws ExecutionException, InterruptedException {
        testProposal(true);
    }

    @Test
    public void proposalFlowCreatesTheCorrectProposalsInBothNodesVaultsWhenInitiatorIsSeller() throws ExecutionException, InterruptedException {
        testProposal(false);
    }


    private void testProposal(Boolean isBuyer) throws ExecutionException, InterruptedException {
        int amount = 1;
        Party counterparty = b.getInfo().getLegalIdentitiesAndCerts().get(0).getParty();
        nodeACreatesProposal(isBuyer, amount, counterparty);
        ImmutableList.of(a,b).forEach(node -> {
            node.transaction(() -> {
                List<StateAndRef<ProposalState>> proposals = node.getServices().getVaultService().queryBy(ProposalState.class).getStates();
                Assert.assertEquals(1, proposals.size());

                ProposalState proposal = proposals.get(0).getState().getData();

                Assert.assertEquals(amount, proposal.getAmount());
                Party buyer;
                Party seller;

                if(isBuyer){
                    buyer = a.getInfo().getLegalIdentitiesAndCerts().get(0).getParty();
                    seller = b.getInfo().getLegalIdentitiesAndCerts().get(0).getParty();
                }else{
                    seller = a.getInfo().getLegalIdentitiesAndCerts().get(0).getParty();
                    buyer = b.getInfo().getLegalIdentitiesAndCerts().get(0).getParty();
                }

                Assert.assertEquals(buyer, proposal.getBuyer());
                Assert.assertEquals(seller, proposal.getSeller());
                return null;
            });
            return;
        });
    }
}

File: negotiation-cordapp/workflows-kotlin/src/test/kotlin/negotiation/flows/flows/ProposalFlowTests.kt
package negotiation.flows.flows

import negotiation.states.ProposalState
import net.corda.core.node.services.queryBy
import net.corda.testing.internal.chooseIdentity
import org.junit.Test
import kotlin.test.assertEquals

class ProposalFlowTests: FlowTestsBase() {

    @Test
    fun `proposal flow creates the correct proposals in both nodes' vaults when initiator is buyer`() {
        testProposal(true)
    }

    @Test
    fun `proposal flow creates the correct proposals in both nodes' vaults when initiator is seller`() {
        testProposal(false)
    }

    private fun testProposal(isBuyer: Boolean) {
        val amount = 1
        val counterparty = b.info.chooseIdentity()

        nodeACreatesProposal(isBuyer, amount, counterparty)

        for (node in listOf(a, b)) {
            node.transaction {
                val proposals = node.services.vaultService.queryBy<ProposalState>().states
                assertEquals(1, proposals.size)
                val proposal = proposals.single().state.data

                assertEquals(amount, proposal.amount)
                val (buyer, proposer, seller, proposee) = when {
                    isBuyer -> listOf(a.info.chooseIdentity(), a.info.chooseIdentity(), b.info.chooseIdentity(), b.info.chooseIdentity())
                    else -> listOf(b.info.chooseIdentity(), a.info.chooseIdentity(), a.info.chooseIdentity(), b.info.chooseIdentity())
                }

                assertEquals(buyer, proposal.buyer)
                assertEquals(proposer, proposal.proposer)
                assertEquals(seller, proposal.seller)
                assertEquals(proposee, proposal.proposee)
            }
        }
    }
}